rfcs/0005-version-matrix.md

```diff
@@ -0,0 +1,477 @@
+
+### `Cargo.lock` is committed to source control
+
+Having a one true `Cargo.lock` solves a lot of the contributor and CI issues.
+Contributors can easily use the same dependency versions as the CI does.
+The CI won't suddenly start producing errors just because a new dependency version was published that causes issues.
+We get reproducible historic builds.
+We are in control of the timing when dependency updates are introduced.
+
+### `Cargo.lock` is updated regularly
+
+Given that a `Cargo.lock` of a library is used only by the developers of said library, we should not get comfortable with one set of versions for too long.
+`Cargo.lock` should be updated to contain the most recent versions of dependencies on a regular basis.
+Roughly as often as we update the stable Rust toolchain pin in the CI script.
+Waiting longer than that creates too large of a time window for bad user experience issues to remain hidden.
+Updating it too often, e.g. every week just because, will cut down on the benefits of contributor's local environment matching the CI and increase merge conflicts.
+Critically, an update should always happen just before publishing to ensure that we test the unlocked experience that users will have.
+
+### `Cargo.toml` matches `Cargo.lock`
+
+We don't want to claim that our project works with dependency versions that we don't actually test.
+Thus we want to make sure that our `Cargo.toml` has the minimum versions specified exactly as we test them with `Cargo.lock`.
+
+An easy way to do this is to use the `cargo upgrade` tool available via [`cargo-edit`](https://crates.io/crates/cargo-edit).
+
+```sh
+cargo install cargo-edit
+```
+
+Running the following command will update all the `Cargo.toml` files in the workspace and also `Cargo.lock`, in a SemVer compatible way.
+
+```sh
+cargo upgrade
+```
+
+### `rust-version` (MSRV) is a formalized entry of de facto reality
+
+Instead of striving for a specific MSRV target we just specify the oldest Rust toolchain that is capable of dealing with our code and its dependency tree.
+Projects with few dependencies like Kurbo can continue to enjoy a relatively old MSRV, while larger projects like Vello will specify a much newer MSRV due to the fast moving dependencies demanding it.
+
+For now the only way to know when this needs to be bumped is by failing a compilation attempt with the MSRV toolchain, either locally or via CI.
+In the future Cargo will receive additional powers to inform us more directly under some circumstances.
+
+Additionally, we will accept reasonable PRs even if they contain code which directly uses Rust features that require the MSRV to be bumped all the way to stable.
+What exactly *reasonable* means will be left undefined by this RFC and will remain at the discretion of reviewers, as has been the case thus far.
+
+### MSRV changes are non-breaking
+
+We join the larger Rust community in their choice of declaring MSRV changes as non-breaking.
```

Comment:
Given how frequently we release breaking changes anyway, I think we ought to consider MSRV changes breaking. It seems to me that this would have low cost and be a much better experience for users.

Reply:
Many in the larger Rust ecosystem don't agree with this. As I explained in the RFC it only takes one sub-dependency to bump the MSRV of the whole project. Large Rust projects will inevitably decay to a state where all their published versions have an incorrect `rust-version` field in the eyes of the current Cargo resolver.

Even if we ignore the sub-dependency issues for the moment and just consider our own code, the situation isn't as clear cut as you suggest.

First, most users won't be running old toolchains. To quote yourself from a sibling comment:
> Most people do update almost as soon as a new rust version is release (< 1 week, same day is common)

Second, the experience for the old toolchain users will be indeed better in terms of compilation not breaking after `cargo update`. Although even that is scheduled to change with Cargo resolver v3. However even today they can either revert their `Cargo.lock` changes, or use `--precise` to specifically downgrade the problematic packages. Keep in mind that in today's Rust climate people with old toolchains need to manually craft their `Cargo.lock` all the time anyway. It's why there is so much work being done by the Cargo team with things like [rfc#3537](https://github.com/rust-lang/rfcs/pull/3537).

However compilation success/failure is only part of the user experience. There's also the ease of gaining knowledge about and access to new features and bug fixes. If MSRV changes are breaking, then simple bugfix patch releases suddenly get promoted to breaking releases. That means they aren't received nor learned about via `cargo update`.

So the trade-off with having MSRV changes be always breaking is that we improve the compilation experience of a small minority of users still using the affected toolchains, at the cost of them not even learning about our patch releases via `cargo update`. Plus the bigger cost of most of our users, on qualifying toolchain versions, not receiving the updates either.

While if you flip that and don't consider MSRV changes breaking, well then most users have a smooth experience and receive patches automatically via `cargo update`. Even the people on too old toolchains will learn of the patches via `cargo update`, after which they can decide if they want to revert the `Cargo.lock` changes or update their toolchain.

To me there's no question that considering MSRV not breaking gives our users as a whole more value. Then you add on top of that the fact that our sub-dependencies will be bumping MSRV willy-nilly anyway. The choice crystallises even more in favor of not treating MSRV as breaking.